### ISSUE TITLE
Excluding / doesn't work with --patterns-from

### ISSUE THREAD

**Issue by edgimar** [borgbackup/borg#2314]
In issue #1406 the patterns file format which was chosen should allow something like the following to work, if I'm not mistaken:

    + /home/foo/bar/**
    + /etc/**
    - /**

This should backup the contents of /home/foo/bar and /etc, and nothing else on the filesystem.  Presently using a patterns file like this results in no files getting selected at all.  Regardless of whether it make sense to have a patterns file like this (some might argue that two roots should be specified on the command-line, but there are use-cases when it is preferable to have just a single root of /), the above should still work.

For the record, I am using borg (1.1.0b4.dev223+ng10d4c97c) from the command-line as follows:

    borg-backup create --exclude-if-present .nobackup --keep-exclude-tags --patterns-from patt_file backupserver:/home/backupuser/backups/borgrepo::myarchive /

Backing up without the `--patterns-from` option works fine.

**Comment by edgimar** [comment]
One other note:  I realize now that I can to some extent work around this problem by just specifying roots within the patterns file instead of on the command-line.  But I still do consider the reported behavior (no files getting backed up) to be a bug.

**Comment by edgimar** [comment]
Oops, I take that back -- it really can't be worked around in a reasonable way using roots -- there is a bug that prevents standard use-cases.  For example, say I have a directory tree (beneath a home directory) that looks like:

    mydir/
    mydir/x
    mydir/x/a
    mydir/x/b
    mydir/y

Now let's say I want to backup everything under `mydir`, except that within the `x` subdirectory, I only want to back up `a`, and no other folders within `x`.  So, I attempt to get this behavior with the following pattern file:

    R /home/user/mydir
    + /home/user/mydir/x/y/**
    - /home/user/mydir/x/**

Unfortunately with this patterns file, and using the command-line below, `mydir/x` does not get backed up at all.  If it were working correctly, `mydir/x/y` should get backed up (and nothing else besides that within `mydir/x`).

Command line used:

    borg-backup create --exclude-if-present .nobackup --keep-exclude-tags --patterns-from patt_file backupserver:/home/backupuser/backups/borgrepo::myarchive

(this is the same as in the original issue, except that it doesn't specify a root on the commandline)

**Comment by ThomasWaldmann** [comment]
I guess for a sane handling of this, we need to push the decision about recursion or not to the user.

So it could be something like:
```
+ include/this
- exclude, but recurse into, maybe finding includes
! exclude, do not recurse into
```

**Comment by enkore** [comment]
That sounds about right in my ears. Maybe swap -/! meanings, as --exclude (~ `-`) should not be changed to "exclude, but recurse into". And ! = [danger zone](https://www.youtube.com/watch?v=d3D7Y_ycSms) -- or in this case "slow zone".

**Comment by ThomasWaldmann** [comment]
Well, I guess some users will just use + and - and expect that it'll work correct under all circumstances. if - does not recurse, it won't.

But also, being inconsistent with --exclude isn't nice....

**Comment by edgimar** [comment]
If it matters, using + and - with recursion is also consistent with rdiff-backup's `--exclude-filelist` syntax.

**Comment by enkore** [comment]
Can we get some votes in here please?

Thumbs up here for

```
+ include and recurse into
- exclude, but recurse into, searching for includes
! exclude, do not recurse into
```

**Comment by enkore** [comment]
And thumbs up here for

```
+ include and recurse into
- exclude, do not recurse into
! exclude, but recurse into, searching for includes
```

**Comment by enkore** [comment]
Since that then is just a matter of swapping the two characters later if necessary, this can already be implemented.

**Comment by W0nderer** [comment]
The example given by edgimar at the top would not include anything using rsync either.  It's similar to the non-working example given in the man pages:
```
+ /some/path/this-file-will-not-be-found
+ /file-is-included
- *
```
which has to be re-factored as:
```
+ /some/ 
+ /some/path/ 
+ /some/path/this-file-is-found 
+ /file-also-included 
- * 
```
Unfortunately this wouldn't work in Borg because it appends a "*" to every rule ending with a "/" (issue [#2399](https://github.com/borgbackup/borg/issues/2399)).

I would have voted for "-" meaning "exclude, do not recurse into" because that's the way it works in rsync (if I understand it correctly).

In the "exclude but recurse" rule what happens to any included files or subdirectories below the excluded directory?  Does their presence force the excluded directory to also be included after all, or do they become 'orphans' and are are adopted by the parent of the excluded folder, e.g. `/grand/parent/child` => `/grand/child`.  The first approach would probably act like "exclude folder if empty", the second would change the directory structure of the system being backed up (probably not what users expect!).

**Comment by edgimar** [comment]
Regarding your question about what happens when there are included files below an excluded directory, say you have a directory `/x` which is excluded, however `/x/y/z` is included -- in this case, `/x/y/z` should be backed up (preserving its full path).

Though I think you're right about how rsync interprets include/exclude rules, I find this behavior by rsync to be counterintuitive and leading to weird "hacks" to try to get what you want (like the second 're-factored' example you gave).  It would be preferable (and no less expressive) to be able to accomplish this with:

    + /some/path/this-file-is-found
    - **

in the case that you only want to include one specific file.  On the other hand, if you want to include all files beneath a path except for a specific file, you could have something like:

    - /some/path/this-file-is-excluded
    + /some/path/
    - **

The point is that reasoning about what will or won't be included is much easier if you can think of borg simply stepping through the list until it finds the first match.  Apparently this is not how rsync works.

**Comment by W0nderer** [comment]
Thanks for clarifying how the "exclude but recurse" rule works, so the effect is to exclude a directory unless it has any descendants to be included.

I agree that it will be better if Borg can filter directories and files in the way you say, without having to match all the ancestor directories as given in rsync's re-factored example, but aren't sure how it achieves this.

According to rsync man pages:

> As the list of files/directories to transfer is built, rsync checks each name to be transferred against the list of include/exclude patterns in turn, and the first matching pattern is acted on: if it is an exclude pattern, then that file is skipped; if it is an include pattern then that filename is not skipped; if no matching pattern is found, then the filename is not skipped.

and

> ... every subcomponent of every path is visited from the top down, so include/exclude patterns get applied recursively to each subcomponent’s full name (e.g. to include "/foo/bar/baz" the subcomponents "/foo" and "/foo/bar" must not be excluded). The exclude patterns actually short-circuit the directory traversal stage when rsync finds the files to send.

Does Borg achieve similar results in a similar way, or use a completely different approach?

A breadth-first versus depth-first traversal shouldn't make any difference to the results should it, provided there aren't any side effects en-route that affect later nodes.  In a depth-first traversal when a directory has been excluded the rscnc can immediately skip to the next sibling rather than recursing down the descendants.  In a breadth-first traversal it will have to mark the directory as excluded so that it is skipped when the next level down is traversed.  Either way all the descendants will be excluded, but no directories or files in other parts of the tree will be affected will they?  Perhaps I am missing something here, do you have any examples where it would make a difference?

In verbose mode rsync lists the directories and files being processed in depth first order, so I assumed that the tree is traversed in that order, but it could have created a list in breadth-first order then processed it in depth first.

**Comment by edgimar** [comment]
Sorry, I was wrong about the order rsync traverses the tree, so forget that -- both borg and rsync traverse depth first.  The issue is simply that rsync won't recurse into directories that have been excluded to see if there are any children that are included, whereas borg does do this (though the rsync-like behavior can be gotten by using `!` instead of `-` at the moment with PR #2322).

**Comment by leo-b** [comment]
Sorry, I missed this discussion and it seems to be already closed. :-( But nevertheless I'd like to add some comments...

Even with the original ``--patterns-from`` implementation, this should have been possible (similar to rsync and the approach @W0nderer suggested but without the trailing slashes):

```
+ /some
+ /some/path
+ /some/path/this-file-is-found 
+ /file-also-included 
- /**
```

And I'd also vote against changing the behavior of ``-`` to recurse into. This will lead to bad performance and unexpected delays, e.g. when remote mount points or something like ``/proc`` or ``/sys`` is anywhere below ``- ...`.

**Comment by W0nderer** [comment]
This approach could be made easier by a new extended type of `+` rule prefix (e.g. `%`) that tries to match all parent directories first (analogous  to `mkdir --parents`) so that:
```
% /some/path/this-file-is-found
```
is equivalent to: 
```
+ /some
+ /some/path
+ /some/path/this-file-is-found
```
But could I also request that `%` does NOT automatically add a `*` after a trailing `/`, as raised in issue [#2399](https://github.com/borgbackup/borg/issues/2399).  Not being able to use patterns ending in `/` to match only directories (as in `rsync` and `bash`) is a serious limitation in some cases.  IMHO `+` should also work in this way.

**Comment by leo-b** [comment]
> This approach could be made easier by a new extended type of + rule prefix (e.g. %) that tries to match all parent directories first.

I like that idea!

> But could I also request that % does NOT automatically add a * after a trailing /, as raised in issue #2399. 

Agreed. This implicit adding of ``*`` is really cumbersome.
And yes, a syntax for matching only directories would also be great.
So changing the behavior of ``path/`` to match only directories would be the logical consequence.
However, this would be a breaking change, so I doubt that we'll get the borg maintainers on board... ;-)

**Comment by W0nderer** [comment]
> However, this would be a breaking change, so I doubt that we'll get the borg maintainers on board... ;-)

Having existing parameters such as `--exclude-from` change the way they work within major releases would be highly undesirable for released versions being used in production environments.  However as far as I know `--patterns-from` is only available in version 1.1, and that hasn't been released yet, so perhaps it's not too late!  The [Compatibility notes](https://borgbackup.readthedocs.io/en/stable/) say "NOT RELEASED DEVELOPMENT VERSIONS HAVE UNKNOWN COMPATIBILITY PROPERTIES.".
